src/QsCompiler/DataStructures/RuntimeCapability.fs

```diff
@@ -3,76 +3,139 @@
-
-    /// Returns true if having this runtime capability also implies having the given runtime capability.
-    member x.Implies y =
-        match x, y with
-        | BasicMeasurementFeedback, BasicQuantumFunctionality
-        | FullComputation, _ -> true
-        | _ -> x = y
-
-    /// Returns a runtime capability that implies both given capabilities.
-    static member Combine x y =
-        match x, y with
-        | BasicQuantumFunctionality, other
-        | other, BasicQuantumFunctionality -> other
-        | BasicMeasurementFeedback, BasicMeasurementFeedback -> BasicMeasurementFeedback
-        | FullComputation, _
-        | _, FullComputation -> FullComputation
-
-    /// The base runtime capability is the identity element when combined with another capability. It is implied by
-    /// every other capability.
-    static member Base = BasicQuantumFunctionality
-
-    /// Returns true if both runtime capabilities are equal.
-    static member op_Equality(a: RuntimeCapability, b: RuntimeCapability) = a = b
-
-    /// Parses the string as a runtime capability.
-    static member TryParse value =
-        // TODO: RELEASE 2021-04: Remove parsing for "QPRGen0", "QPRGen1", and "Unknown".
-        match value with
-        | "BasicQuantumFunctionality"
-        | "QPRGen0" -> Value BasicQuantumFunctionality
-        | "BasicMeasurementFeedback"
-        | "QPRGen1" -> Value BasicMeasurementFeedback
-        | "FullComputation"
-        | "Unknown" -> Value FullComputation
-        | _ -> Null
-
-    member this.Name =
-        match this with
-        | BasicQuantumFunctionality -> "BasicQuantumFunctionality"
-        | BasicMeasurementFeedback -> "BasicMeasurementFeedback"
-        | FullComputation -> "FullComputation"
-
-// TODO: RELEASE 2021-04: Remove RuntimeCapabilitiesExtensions.
-[<Extension>]
-[<Obsolete "Use Microsoft.Quantum.QsCompiler.RuntimeCapability.">]
-module RuntimeCapabilitiesExtensions =
-    [<Extension>]
-    [<Obsolete "Use Microsoft.Quantum.QsCompiler.RuntimeCapability.">]
-    let ToCapability capabilities =
-        match capabilities with
-        | RuntimeCapabilities.QPRGen0 -> BasicQuantumFunctionality
-        | RuntimeCapabilities.QPRGen1 -> BasicMeasurementFeedback
-        | _ -> FullComputation
+    {
+        resultOpacity: ResultOpacity
+        classical: ClassicalCapability
+    }
+
+    member capability.ResultOpacity = capability.resultOpacity
+
+    member capability.Classical = capability.classical
+
+    static member BasicExecution = { resultOpacity = Opaque; classical = Empty }
+
+    static member AdaptiveExecution = { resultOpacity = Transparent; classical = Integral }
+
+    static member BasicQuantumFunctionality = { resultOpacity = Opaque; classical = Full }
+
+    static member BasicMeasurementFeedback = { resultOpacity = Controlled; classical = Full }
+
+    static member FullComputation = { resultOpacity = Transparent; classical = Full }
+
+module RuntimeCapability =
```

Comment:
Discussed during PR walkthrough: Since we make a bunch of breaking changes, we should just take the opportunity to align terminology across the board and call this TargetCapability.

Reply:
I renamed it, but there are still leftovers:

- The qsc command line option is called `--runtime` but should probably be `--target-capability` or `--capability`.
- The project property is still called `ResolvedRuntimeCapabilities` - this needs an SDK change.

Reply:
I can go an update that if you merge the changes here into the hardware targeting branch. 
Is this PR on its own fully functional? If so, I suggest I just cover those as part of the second half (getting the hardware_targeting branch merged) that contains the Sdk changes.

Reply:
Yeah, everything should work with this PR, just the names will be inconsistent. I'm fine with doing it in another PR.